fedimint-server/src/consensus/engine.rs

```diff
@@ -804,3 +1019,44 @@ pub async fn get_finished_session_count_static(dbtx: &mut DatabaseTransaction<'_
         .await
         .map_or(0, |entry| (entry.0 .0) + 1)
 }
+
+/// In memory copy of all the current consensus versions that are maintained in
+/// the DB.
+///
+/// To avoid reading from the database all the time, invalidation is supported.
+/// Any time consensus item that could have changed it is processed, the whole
+/// thing will get invalidated and reloaded at the beginning for processing new
+/// consensus item.
+///
+/// Create by using [`ConsensusEngine::load_consensus_version_cache`]
+#[derive(Debug)]
+struct ConsensusVersionCache {
+    // Currently nothing is using it, but in the future we can pass these
+    // to modules.
+    #[allow(unused)]
+    versions: BTreeMap<Option<ModuleInstanceId>, ConsensusVersion>,
+    invalidated: bool,
+}
```

Comment:
Caching this is way to risky.

Reply:
@joschisan

I agree that caching here is a potential source of bugs, but I don't see any practical implementation without it.

Since we need to pass the consensus version for a given module to every `process_x` we can't be reading it from the db every time. That would possibly x10 amount of db reads in consensus processing. 


I implemented the caching carefully: there's only single place where the actual versions could change (processing of a vote), and the cache gets invalidated whenever that code runs.